### PR TITLE
Keras vectorizer build_embedding_matrix additions

### PR DESCRIPTION
**Pull request by lizgzil**
Description
---

Fixing https://github.com/wellcometrust/WellcomeML/issues/35

I can't actually see any usage of the "build_embedding_matrix" function in other projects, so I'm hoping setting `embeddings_path=None` doesn't break anything.

I've set it so that the function can take a gensim word vector name if the local path isn't given. But can change the logic if this isn't ideal, I wasn't sure whether it'd be better to just give a default model name?

Checklist
---

- [ ] Added link to Github issue or Trello card
- [ ] Added tests

### CODE REVIEW

File: wellcomeml/ml/keras_vectorizer.py [hunk 1]
```diff
@@ -21,20 +26,58 @@ def transform(self, X, *_):
...
+            if path.exists(embeddings_path):
+                embeddings_index = {}
+                with open(embeddings_path) as f:
+                    for line in f:
+                        word, coefs = line.split(maxsplit=1)
+                        coefs = np.fromstring(coefs, "f", sep=" ")
+                        embeddings_index[word] = coefs
+                    emb_dim = len(coefs)
+            else:
+                logger.error("Incorrect local embeddings path")
```
Comment (nsorros) [inline wellcomeml/ml/keras_vectorizer.py:54]: should we log an error or throw an exception. does the pipeline work if you get None back?

File: wellcomeml/ml/keras_vectorizer.py [hunk 2]
```diff
@@ -21,20 +26,58 @@ def transform(self, X, *_):
...
         return pad_sequences(sequences, maxlen=self.sequence_length)
 
-    def build_embedding_matrix(self, embeddings_path):
-        embeddings_index = {}
-        with open(embeddings_path) as f:
-            for line in f:
-                word, coefs = line.split(maxsplit=1)
-                coefs = np.fromstring(coefs, "f", sep=" ")
-                embeddings_index[word] = coefs
+    def build_embedding_matrix(self, embeddings_path=None, word_vectors=None):
```
Comment (nsorros) [inline wellcomeml/ml/keras_vectorizer.py:29]: i wonder whether the two arguments should be one, check the `from_pretrained` argument in transformers https://huggingface.co/transformers/main_classes/model.html#transformers.PreTrainedModel.from_pretrained.

so the idea would be that the user passes either a path to embeddings stored locally or the name of a pretrained embedding. those embeddings are downloaded and cached so second time do not need to download again

**Review by nsorros — CHANGES_REQUESTED** [review]
That is a great addition @lizgzil, there is a discussion point on whether we want to have one argument that received either a path or a name and if it receives a name it downloads and caches the embeddings. I think this would be a nice approach that follows ideas elsewhere in our library as well.

File: wellcomeml/ml/keras_vectorizer.py (hunk 1 above)
Comment (lizgzil) [reply wellcomeml/ml/keras_vectorizer.py:54]: yeh i think youre right it should really throw an exception and stop

**Comment by lizgzil** [comment]
@nsorros I've made those changes. It looks like when you download a gensim model it will cache it automatically and get the file from there going forward - so didn't need to write anything for this. 

"Gensim has a gensim.downloader module for programmatically accessing this data. The module leverages a local cache that ensures data is downloaded at most once." https://radimrehurek.com/gensim/auto_examples/howtos/run_downloader_api.html

File: wellcomeml/ml/keras_vectorizer.py [hunk 3]
```diff
@@ -21,20 +24,59 @@ def transform(self, X, *_):
...
+        else:
+            raise TypeError("No local or GenSim word embeddings given")
 
-        emb_dim = len(coefs)
         num_words = len(self.tokenizer.word_index) + 1
 
         embedding_matrix = np.zeros((num_words, emb_dim))
         for word, i in self.tokenizer.word_index.items():
-            embedding_vector = embeddings_index.get(word)
+            if local_embeddings:
```
Comment (nsorros) [inline wellcomeml/ml/keras_vectorizer.py:72]: is that needed, why not use the try except block for both cases?

**Review by nsorros — APPROVED** [review]
LGTM, one possible consideration.

File: wellcomeml/ml/keras_vectorizer.py (hunk 3 above)
Comment (lizgzil) [reply wellcomeml/ml/keras_vectorizer.py:72]: I think it is because the `get` command wont work for the gensim embeddings case

Comment (lizgzil) [reply wellcomeml/ml/keras_vectorizer.py:72]: I think it is because the `get` command wont work for the gensim embeddings case

**Pull request merged**